doc/dev/architecture.md

```diff
@@ -0,0 +1,304 @@
+
+- **Compiler Core:** An MLIR-based compiler for hybrid quantum programs.
+    Provides optimizations and other transformations such as automatic differentiation, with a
+    growing library of compilation passes.
+    Targets LLVM IR with QIR syntax for code generation.
+
+- **Runtime:** A runtime library for execution of hybrid quantum programs.
+    While classical computation is compiled to native code, all quantum functionality is managed and
+    provided by the runtime as an interface layer to backend devices (such as CPU simulators,
+    GPU simulators, and hardware).
+
+
+## Frontend
+
+An overview of the Python frontend is presented below. At the moment, the frontend is not only
+resposible for converting user programs to the compiler IR, but also comes with a compiler driver
+that manages the entire compilation pipeline.
+
+![img](../_static/arch/frontend.svg)
+
+Generally speaking, compilation happens in 3 stages which are successively invoked by the frontend
+and compiler driver:
+
+- **Program Capture / IR Generation:** The frontend primarily provides a method for hybrid program
+    capture of PennyLane/JAX programs. This is achieved by leveraging the tracing & op queueing
+    mechanism of both frameworks, extending the JAX program representation (JAXPR) with quantum
+    primitives. Custom JAXPR -> MLIR lowerings are registered to these primitives to fully convert
+    a hybrid program to MLIR for consumption by the compiler.
+
+- **Program Transformation:** The main phase of the compilation process is performed on the
+    MLIR-based representation for hybrid quantum programs defined by Catalyst.
+    The user program is passed to the Catalyst compiler libraries in its textual form, as the MLIR
+    memory objects are not compatible between Catalyst and `jaxlib`.
+    The driver then invokes a sequence of transformations that lowers the user program to a lower
+    level of abstraction, outputing LLVM IR with QIR syntax.
+    For more details consult the [next section](#compiler-core).
+
+- **Code Generation:** At this stage the LLVM IR is compiled down to native object code using the
+    LLVM Static Compiler (`llc`) for the local system architecture. A native linker is then used
+    to link the user program to the Catalyst Runtime library.
+    The frontend will load this library into the Python environment and attach its entry point to
+    the callable `@qjit` object defined by the user.
+    For more details on the program execution consult the [runtime section](#runtime--execution).
+
+Elaborating on the program capture phase, tracing is a mechanism by which a function is executed
+with abstract arguments called *tracers*. Calling operations from the tracing library (such as
+`jax.numpy`) will record the operation and which tracers (or constants) it acts upon in a global
+data structure. Note that JAX's tracing contexts are thread-local and can be nested to allow scoped
```

Comment:
No this is what I meant, I wanted to point out that the tracing data structures are global but local to each thread, so tracing of different functions can be performed in parallel without interfering with each other. This is not the case in PennyLane per a recent discussion with Josh, but I suppose it could be considered an unnecessary detail.

Reply:
Well, this details look unnecessary to me in this context. Just my opinion, I am not sure do we need to change it or not.

Reply:
Yes you are probably right, I think I added it because it was just fresh on my mind 😅 Will remove

Reply:
Removed!